File: 2021/6/6.jl
#!/usr/bin/env julia
# usage: julia 6.jl [FILENAME]

module AOC2021Day6

load(file::AbstractString)::Vector{Int} = parse.(Int, split(readchomp(file), ','))

descendants = let
    cache = Dict{Int,Int}()
    function (age)
        get!(cache, age) do
            kids = age-9:-7:0
            sum(descendants, kids; init=length(kids))
        end
    end
end

calculate(fish, days) = length(fish) + sum(descendants, days + 8 .- fish)

function calculatematrix(fish, age)
    F = [count(==(n), fish) for n in 0:8]
    P = [
        0 1 0 0 0 0 0 0 0
        0 0 1 0 0 0 0 0 0
        0 0 0 1 0 0 0 0 0
        0 0 0 0 1 0 0 0 0
        0 0 0 0 0 1 0 0 0
        0 0 0 0 0 0 1 0 0
        1 0 0 0 0 0 0 1 0
        0 0 0 0 0 0 0 0 1
        1 0 0 0 0 0 0 0 0
    ]
    sum(P^age * F)
end

partone(fish) = calculate(fish, 80)
parttwo(fish) = calculate(fish, 256)

if abspath(PROGRAM_FILE) == @__FILE__
    fish = load(get(ARGS, 1, joinpath(@__DIR__, "input.txt")))
    println(partone(fish))
    println(parttwo(fish))
elseif isinteractive()
    using REPL
    REPL.activate(AOC2021Day6)
end

end # module

File: 2021/6/6.pl
#!/usr/bin/env swipl
% usage: swipl 6.pl [FILENAME]

:- module(aoc2021day6, []).
:- module(aoc2021day6).
:- initialization(main, main).
:- use_module(library(clpfd)).
:- use_module(library(dcg/basics)).
:- use_module(library(dcg/high_order)).
:- use_module(library(solution_sequences)).

load(Result) --> sequence(integer, ",", Fish0), eol, eos,
    {
        msort(Fish0, Fish1),
        clumped(Fish1, Clumps),
        foldl([Fish - Count, R0, R] >> (
            nth0(Fish, R0, _, R1),
            nth0(Fish, R, Count, R1)
        ), Clumps, [0, 0, 0, 0, 0, 0, 0, 0, 0], Result)
    }.

population_model(Before, After) :-
    Before = [F0, F1, F2, F3, F4, F5, F6, F7, F8],
    Next   = [F1, F2, F3, F4, F5, F6, F,  F8, F0],
    F #= F0 + F7,
    ( After = Next ; population_model(Next, After) ).

population_after_n_days(N, Fish, Population) :-
    call_nth(population_model(Fish, FishN), N),
    sum_list(FishN, Population).

part_one(Fish, Population) :- population_after_n_days( 80, Fish, Population).
part_two(Fish, Population) :- population_after_n_days(256, Fish, Population).

main([File]) :-
    once(phrase_from_file(load(Fish), File)),
    part_one(Fish, ResultOne),
    writeln(ResultOne),
    part_two(Fish, ResultTwo),
    writeln(ResultTwo).

main([]) :- main(['input.txt']).
